pkg/kubelet/network/cni/cni_windows.go

```diff
@@ -0,0 +1,64 @@
+// +build windows
+
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cni
+
+import (
+	"fmt"
+	cniTypes020 "github.com/containernetworking/cni/pkg/types/020"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/network"
+)
+
+func (plugin *cniNetworkPlugin) Init(host network.Host, hairpinMode kubeletconfig.HairpinMode, nonMasqueradeCIDR string, mtu int) error {
+
+	plugin.host = host
+
+	plugin.syncNetworkConfig()
+	return nil
+}
+
+func getLoNetwork(binDir, vendorDirPrefix string) *cniNetwork {
+	return nil
+}
+
+// GetPodNetworkStatus : Assuming addToNetwork is idempotent, we can call this API as many times as required to get the IPAddress
```

Comment:
`addToNetwork` is definitely not idempotent. The CNI spec doesn't require it and no plugins that I'm aware of implement ADD in this manner. The comment from before is... optimistic.

We're currently working on a spec change for a GET method within CNI.

Reply:
But it's a plugin system. You cannot guarantee that users will use your plugin.

Reply:
I agree with squeed here.

Reply:
Well, let me put it this way. In windows, There is no other way or tools like nsenter available to get the IPAddress of the endpoint attached to the container in a namespace, other than calling the CNI Plugin again. The concepts are different for Windows. It would be better for the Plugin which allocated an endpoint, to get back the IPAddress of the endpoint it manages. It may not be feasible to find some alternative ways to get the endpoint ipaddress, like trying to use some other inbuilt tools. There are other complications there. 
Hence it is made idempotent here. This is a requirement for Windows CNI Plugins.
Anyone writing a windows CNI plugin, has to know this platform requirement, which would be published separately, somewhere. 

There is another requirement of windows CNI plugin, which is different from the Linux CNI plugin, like the CNI plugin has to be invoked with ADD for second or subsequent containers created on a POD, with netNsName as the "container:<Infra POD ContainerID>". This is again a platform requirement to replicate some of the critical data to the new container. Comment for this is added in the code itself.

Plugins are written this way to that platform specific requirements could be offloaded to the Plugin rather than putting them in the Kubernetes code, following the CNI spec.

Reply:
If the plugin can "enter the namespace and list the IPs," then why can't Kubernetes?

Reply:
The Plugin doesn't enter the namespace and get the IP directly. It calls into Host Networking Service, to get the data. Kubernetes can also do that directly, but then we would destroy the very idea of abstracting that logic out behind the plugin and not bringing in a lot of platform specific APIs.

The main reason behind this approach is that, current containers are launched with process isolation. It could also be launched with hyper-v isolation (not supported in kubernetes but supported by docker and platform). To keep the implementation same across these modes, it is better to call the centralized to CNI plugin to get back the data, the current API would seamlessly support these.

Reply:
@madhanrm I agree with @squeed.  The CNI specification does not declare ADD idempotent.  That has nothing to do with the platform at all, that's just how it is.  Making Windows act differently is simply confusing.  So can we discuss this a bit more so that we all understand the restrictions?

Saying "This is a requirement for Windows CNI Plugins." is a solution to a problem, but there may be other solutions that do not require making ADD idempotent for Windows CNI.

IIRC Windows has no shared "infra" (eg net namespace) container.  Instead, *every* container gets network setup called on it.  What I think we're arguing about here, and correct me if I'm wrong, is that in the current PR here we'll get an ADD for the same container/namespace/netns tuple multiple times, which would in fact require idempotency that CNI does not guarantee.

Instead of that, could we figure out some other way to pass the data that your Windows plugin requires to tie the containers together?

One other thing I'm not clear on.  What actual CNI plugin are you working on?  And why couldn't some other 3rd party Windows plugin have a completely different mechanism for tying containers together, that doesn't require ADD to be idempotent?

Reply:
@dcbw Just for my own understanding, what is the argument against making ADD idempotent in the spec?

Reply:
@dcbw 
The current Idempotent nature of ADD is only required because any subsequent ADD is used as a GET (PodStatus) to get the IP. Once we have GET implemented, that will not be a criteria for ADD to be idempotent. (But IMO, it doesn't make sense for ADD not to be idempotent, which is a separate discussion altogether)

Also, your understanding of "every container gets network setup" is getting changed by this PR. This PR is also to start using shared infra container, for which CNI is a requirement.

You should also understand the difference between the CNI plugin in Linux and in Windows. You really do not have to write individual CNI plugins on Windows unlike Linux. Instead CNI plugin is like a proxy to program the Host Networking Service (HNS) on Windows, which takes care of everything. The information is simply constructed and passed on to HNS. So a single CNI plugin would work for most of the network modes that can work independently on a single host. Other may develop similar plugins, but they have to go thro HNS, anything other than that would not be supported in Windows.
The idempotency is required here, because the CNI plugin doenst have to manage anything and simply pass it on to HNS. Else, the CNI would also have to manage redundant data as HNS. This would simplify the requirement of the Windows CNI plugins.

Reply:
Are there any open items here? Can we not relax the idempotent nature of ADD for Windows here?
Am not seeing any progress in this PR.